### PR TITLE
Deploy staging cluster #1935

### PR DESCRIPTION
**Pull request by callummole**
Me and @sgibson91 have been working through #1935.

- added `turing-staging` config. 
- updated `deploy.py` to incorporate `turing-staging`. 
- added `X.staging` IP records into DNS zone `mybinder.turing.ac.uk`. 

We hit a bit of a snag - the deployment mostly works, but the `binder` pods have a `CrashLoopBackOff` error:

```
Readiness probe failed: Get "http://10.244.1.34:8585/versions": dial tcp 10.244.1.34:8585: connect: connection refused
```

Can someone have a look at the new config files and see if we are missing anything? 

Thanks!

### CODE REVIEW

File: config/turing-staging.yaml [hunk 1]
```diff
@@ -0,0 +1,127 @@
+rojectName: turing
```
Comment (minrk) [inline config/turing-staging.yaml:1]:
```suggestion
projectName: turing
```

File: config/turing-staging.yaml [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    hosts:
+      - static-staging.mybinder.turing.ac.uk
+    tls:
+      secretName: turing-staging-static-tls-crt
+
+redirector:
+  redirects:
+    - type: host
+      host:
+        from: docs-mybinder.turing.10.0.0.1.xip.io
```
Comment (minrk) [inline config/turing-staging.yaml:117]: I suspect this redirect isn't needed. Maybe it was for testing?

File: deploy.py [hunk 3]
```diff
@@ -334,7 +339,7 @@ def main():
         # script is running on CI, proceed with auth and helm setup
         if args.cluster == 'ovh':
             setup_auth_ovh(args.release, args.cluster)
-        elif args.cluster == 'turing':
+        elif args.cluster in list(AZURE_RGs.keys()):
```
Comment (minrk) [inline deploy.py:342]: There's a simpler check for whether something is in the keys of a dict:

```suggestion
        elif args.cluster in AZURE_RGs:
```

**Review by minrk — COMMENTED** [review]
All of this looks great to me. I think the only piece missing is adding a matrix entry in  the 'staging-deploy' matrix entry in `.github/workflows/cd.yml`. Then this deployment should actually happen when it's merged.

**Comment by minrk** [comment]
There don't appear to be any DNS records for `staging.mybinder.turing.ac.uk` or any of the subdomains, as far as I can tell:

```
nslookup binder-staging.mybinder.turing.ac.uk
nslookup staging.mybinder.turing.ac.uk
```

File: config/turing-staging.yaml (hunk 2 above)
Comment (sgibson91) [reply config/turing-staging.yaml:117]: Possibly. There's a similar one in the prod config that we just copied over and added `-staging` to

**Comment by callummole** [comment]
I've redeployed to uksouth, since I was doing #2304. 

The A records are now set (they weren't before - I think we must have set them then redeployed).

The binder pod is experienced `CrashLoopBackOff`. With the error from the logs showing:
```
Loading /etc/binderhub/config/values.yaml
Loading extra config: 01-eventlog
[W 220726 08:02:57 _default:283] Authentication failed using Compute Engine authentication due to unavailable metadata server.
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.9/site-packages/binderhub/__main__.py", line 4, in <module>
    main()
  File "/usr/local/lib/python3.9/site-packages/traitlets/config/application.py", line 971, in launch_instance
    app.initialize(argv)
  File "/usr/local/lib/python3.9/site-packages/binderhub/app.py", line 787, in initialize
    self.event_log = EventLog(parent=self)
  File "/usr/local/lib/python3.9/site-packages/binderhub/events.py", line 50, in __init__
    self.handlers = self.handlers_maker(self)
  File "<string>", line 15, in _make_eventsink_handler
  File "/usr/local/lib/python3.9/site-packages/google/cloud/logging/client.py", line 126, in __init__
    super(Client, self).__init__(
  File "/usr/local/lib/python3.9/site-packages/google/cloud/client/__init__.py", line 320, in __init__
    _ClientProjectMixin.__init__(self, project=project, credentials=credentials)
  File "/usr/local/lib/python3.9/site-packages/google/cloud/client/__init__.py", line 268, in __init__
    project = self._determine_default(project)
  File "/usr/local/lib/python3.9/site-packages/google/cloud/client/__init__.py", line 287, in _determine_default
    return _determine_default_project(project)
  File "/usr/local/lib/python3.9/site-packages/google/cloud/_helpers/__init__.py", line 152, in _determine_default_project
    _, project = google.auth.default()
  File "/usr/local/lib/python3.9/site-packages/google/auth/_default.py", line 579, in default
    raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started
```

I note that in `turing.yaml` config there is the block, and a similar one in `ovh.yaml`:
```
  extraVolumes:
    - name: event-secret
      secret:
        secretName: events-archiver-secrets
  extraVolumeMounts:
    - name: event-secret
      mountPath: /event-secret
      readOnly: true
  extraEnv:
    GOOGLE_APPLICATION_CREDENTIALS: /event-secret/service-account.json
```

This is missing in `turing-staging.yaml`. I'm not entirely sure what event secrets are.  I've naively copied this block in its entirety over to `turing-staging.yaml`, but have the same error. Do we need some of this in the `turing-staging.yaml` config?

**Comment by sgibson91** [comment]
Sharing my input on Slack about this for transparency:

> The events archive is hosted in the main GCP project and collect events from all the clusters to create the archive, which then people make cool graphs with regarding launches. It's a centralised service, hence why I think it'll be shared credentials. There is a `secrets/events-archiver/turing.json` file in the repo

**Comment by callummole** [comment]
I am getting a `MountVolume.SetUp failed for volume "event-secret" : secret "events-archiver-secrets" not found` error. I can see the template [here](https://github.com/jupyterhub/mybinder.org-deploy/blob/master/mybinder/templates/events-archiver/secret.yaml), which I assume it wants to use but cannot find for some reason?

**Comment by callummole** [comment]
Some investigations:

- in  `mybinder/values.yml` that there are the following lines:
```
# this is defined in secrets/ for the OVH cluster
eventsArchiver:
  serviceAccountKey: ""
```
- the template `mybinder/templates/events-archiver/secret.yaml` only gets processed if there is something in `events.Archiver.serviceAccountKey, indicated by the first line {{- if .Values.eventsArchiver.serviceAccountKey }}`
also that there are secrets for clusters in `secrets/`. 
- I have assumed it looks automatically for the release name, but couldn't find a secret with that name, so redeployed with `python deploy.py turing-staging turing-staging --name turing`, which would use the `turing-staging` config file but keep the helm release and namespace called `turing`. Same mount error as above.

**Comment by minrk** [comment]
I added a new service account key in #2319 and added it to secrets/config/turing-staging.yaml here. It would really have been fine to copy the same value from `secrets/config/turing.yaml`, but this is perhaps more strictly separated.

**Comment by callummole** [comment]
Thanks! The deploy works (after adding a new node due to maxing cpu). 

I haven't added to the staging fed since the certificates aren't being issued. 

In the certificate challenge the reason for the non-issue is as follows:
```
Reason:      Waiting for HTTP-01 challenge propagation: failed to perform self check GET request 'http://staging.turing.mybinder.org/.well-known/acme-challenge/lTcWYhGTc12klV6v0tDRr3CjprOVQYoF3rHgfl9e-38': Get "http://staging.turing.mybinder.org/.well-known/acme-challenge/lTcWYhGTc12klV6v0tDRr3CjprOVQYoF3rHgfl9e-38": dial tcp: lookup staging.turing.mybinder.org on 10.0.0.10:53: no such host
```

I think I need to do something in azure to link staging.turing.mybinder.org to our IP, but it currently escapes me. @sgibson91 can you remember?

**Comment by sgibson91** [comment]
@callummole You most likely need to setup A records https://alan-turing-institute.github.io/hub23-deploy/advanced/enabling-page-redirection.html#creating-a-records

**Comment by callummole** [comment]
@sgibson91, yes, I have done that for the DNS zone `mybinder.turing.ac.uk`. Each component links to corresponding public IP in the respective cluster
Screenshot 2022-08-08 at 10 37 59.

But I'm not sure how the turing-prod cluster is aware of `turing.mybinder.org`. (and therefore how to link `turing-staging` with `staging.mybinder.org`). I can't see this specified anywhere in the azure A records.

I suspect it is something simple I am overlooking.

**Comment by sgibson91** [comment]
> But I'm not sure how the turing-prod cluster is aware of `turing.mybinder.org`. (and therefore how to link `turing-staging` with `staging.mybinder.org`). I can't see this specified anywhere in the azure resources.

This is set in the mybinder.org DNS zone which we don't have access to. I know @choldgraf owns the account, and there's probably a few other people with access too. But this isn't an Azure thing, the mybinder.org DNS is hosted by bluehost I believe.

**Comment by callummole** [comment]
I guess I could use only `staging.mybinder.turing.ac.uk` instead.

File: config/turing-staging.yaml [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    hosts:
+      - binder-staging.mybinder.turing.ac.uk
+      - staging.turing.mybinder.org
+    annotations:
+      kubernetes.io/ingress.class: nginx
+    tls:
+      - secretName: turing-binder-tls-crt
+        hosts:
+          - binder-staging.mybinder.turing.ac.uk
+          - staging.turing.mybinder.org
```
Comment (sgibson91) [inline config/turing-staging.yaml:29]: For sake of mergeability, we can delete all "staging.turing.mybinder.org" lines from the config. That should remove the cert-manager error you reported earlier. We can add these hosts back in once the CNAME has been added to the mybinder.org DNS zone.

Comment (callummole) [reply config/turing-staging.yaml]: I have done so - see latest commits. 

But does this have implications for the federation redirect? Which is within the `staging.mybinder.org` DNZ zone.

https://github.com/jupyterhub/mybinder.org-deploy/blob/279a9ddb55193b7e898f85b88feb0c64d04ecb0b/config/staging.yaml#L153-L161

Comment (sgibson91) [reply config/turing-staging.yaml]: No that is only indicating which URL is the host of the redirector.

We need to expand the Turing block below that here:

https://github.com/jupyterhub/mybinder.org-deploy/blob/279a9ddb55193b7e898f85b88feb0c64d04ecb0b/config/staging.yaml#L167-L168

It should end up looking like the GKE block below, and we can set the URL to staging.mybinder.turing.ac.uk for now

https://github.com/jupyterhub/mybinder.org-deploy/blob/279a9ddb55193b7e898f85b88feb0c64d04ecb0b/config/staging.yaml#L157-L161

Comment (callummole) [reply config/turing-staging.yaml]: Done - [44e07e2](https://github.com/jupyterhub/mybinder.org-deploy/pull/2170/commits/44e07e2c9dff746d8fd8e48d8e1d605b50695b05)

File: config/staging.yaml [hunk 5]
```diff
@@ -165,4 +165,7 @@ federationRedirect:
     ovh:
       weight: 0
     turing:
-      weight: 0
+      url: https://binder-staging.mybinder.turing.ac.uk
+      weight: 4
```
Comment (sgibson91) [inline config/staging.yaml:169]: We should double check that this weighting is reasonable

File: config/staging.yaml [hunk 6]
```diff
@@ -165,4 +165,7 @@ federationRedirect:
...
+      health: https://binder-staging.mybinder.turing.ac.uk
+      versions: https://binder-staging.mybinder.turing.ac.uk
```
Comment (sgibson91) [inline config/staging.yaml:171]:
```suggestion
      health: https://binder-staging.mybinder.turing.ac.uk/health
      versions: https://binder-staging.mybinder.turing.ac.uk/versions
```

Correcting the URLs for the health and versions endpoints

File: config/turing-staging.yaml [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      # Docker Hub registry for Turing
+      use_registry: true
+      image_prefix: turingmybinder/binder-staging-r2d-g5b5b759-
+
+  replicas: 1
+  ingress:
+    enabled: true
+    hosts:
+      - binder-staging.mybinder.turing.ac.uk
+      # - staging.turing.mybinder.org
```
Comment (sgibson91) [inline config/turing-staging.yaml:22]: Can we add an in-line comment saying that the `staging.turing.mybinder.org` URL can be reintroduced once the CNAME has been setup please?

**Review by sgibson91 — APPROVED** [review]
A few comments, but I believe this is good to go

File: config/staging.yaml (hunk 5 above)
Comment (callummole) [reply config/staging.yaml:169]: @minrk, I just added the same weight as gke staging. Is this appropriate?

Comment (minrk) [reply config/staging.yaml:169]: Yup, that's fine

**Comment by callummole** [comment]
CD continues to be thorny. 

There is intermittent a webhook error (https://github.com/jupyterhub/mybinder.org-deploy/actions/runs/2838356726/attempts/1) which I haven't managed to get to the bottom of. The same happens on `turing-prod`. It may just that sometimes it takes more time.

On a second attempt the deployment works but the test launch fails (https://github.com/jupyterhub/mybinder.org-deploy/actions/runs/2838356726).

The logs show the that the jupyerhub api service is unhealthy. 
```
[E 220811 08:22:41 health:50] Error checking check_jupyterhub_api
    Traceback (most recent call last):
      File "/usr/local/lib/python3.9/site-packages/binderhub/health.py", line 48, in wrapper
        res = await f(*args, **kwargs)
      File "/usr/local/lib/python3.9/site-packages/binderhub/health.py", line 26, in wrapper
        return await f(*args, **kwargs)
      File "/usr/local/lib/python3.9/site-packages/binderhub/health.py", line 140, in check_jupyterhub_api
        await AsyncHTTPClient().fetch(hub_url + "hub/api/health", request_timeout=3)
    tornado.curl_httpclient.CurlError: HTTP 599: Operation timed out after 3001 milliseconds with 0 bytes received
[W 220811 08:22:41 health:208] Unhealthy services: [{'service': 'JupyterHub API', 'ok': False}]
```
Would this interfere with the build?

**Comment by callummole** [comment]
I just checked the `turing-prod` binder pod logs and it has the same jupyerhub-api error, but passes the test launch.

**Pull request merged**